### PR TITLE
Make signing STAC objects more convenient

### LINKED ISSUE #11: `sign_items` convenience method
It's a one-liner, but I find myself rewriting
```python
items = search.items_as_collection()
items = pystac_client.ItemCollection([pc.sign_assets(item) for item in items.features])
```
a lot. Might be nice to have built-in.

Could be even nicer to be able to pass in a Search and have it load and sign the items without making an intermediate ItemCollection?

### PR DESCRIPTION
**Pull request by kshepard**
## Overview

Previously, the library only supported signing a  PySTAC `Item` or an HREF. To make it more convenenient, the `sign` function has been expanded to work with each of: HREF (str), `Asset`, `Item`, and `ItemCollection`. ~In addition, a `search_and_sign` function has been added which performs a search and returns the resulting `ItemCollection` with all assets signed.~

Edit: added a commit to make use of functools singledispatch to allow the `sign` function to also work with `ItemSearch` objects.

## Notes

~Initially I was hoping the `sign` function was going to be able to be applied to the `ItemSearch`, but this isn't possible without loosening the types, because the `sign` function needs to return the same type that is provided. This works in all of the cases except for `ItemSearch`, which needs to instead return an `ItemCollection` with the signed assets.~

Edit: no longer the case, now that functools singledispatch has been added. 

Fixes #11

### CODE REVIEW

File: planetary_computer/sas.py [hunk 1]
```diff
@@ -49,22 +49,37 @@ def ttl(self) -> float:
...
+    if isinstance(obj, Item):
+        return sign_item(obj)
+
+    if isinstance(obj, Asset):
+        return sign_asset(obj)
+
+    if isinstance(obj, ItemCollection):
+        return sign_item_collection(obj)
+
+    raise TypeError("Invalid type, must be one of: str, Asset, Item, or ItemCollection")
```
Comment (TomAugspurger) [inline planetary_computer/sas.py:82]: Thoughts on using [`functools.singledispatch`](https://docs.python.org/3/library/functools.html?highlight=singledispatch#functools.singledispatch) as an alternative to these isinstance checks?

It's 90% a style thing, but in theory 3rd-party code could register an implementation for their own types.

Comment (kshepard) [reply planetary_computer/sas.py:82]: Interesting! I wasn't aware of that, but I'll give it some consideration.

Comment (moradology) [reply planetary_computer/sas.py:82]: New to me as well - as things stand any additional sign implementations will be stuck with that `T->T` shape. That might be just fine (perhaps even desirable?). If the freedom to implement `T->?` is desirable, the use of `functools.singledispatch` seems like the ideal solution

**Review by moradology — APPROVED** [review]

File: planetary_computer/sas.py (hunk 1 above)
Comment (kshepard) [reply planetary_computer/sas.py]: @TomAugspurger I added a commit that switches over to `singledispatch`. I think in general it's a net positive change, so thanks for the recommendation. The only thing I like better about the previous implementation was the better compile time type checking to ensure the function was only accepting one of the supported types. But as @moradology noted, it restricted `sign` to only work on `T->T` signatures, and we've already encountered a need for a divergence with the ability to sign an `ItemSearch` and get back a signed `ItemCollection`. I think the added flexibility and better ease of use for the user (not needing to call a separate function when dealing with an `ItemSearch`) may be a good enough reason to warrant the loosening of types.

Comment (TomAugspurger) [reply planetary_computer/sas.py]: Great, thanks.

In principle I like the idea of having `sign` always return the same type as the input. But as you say, I don't know how that would work for an `ItemSearch`.

**Review by TomAugspurger — APPROVED** [review]

**Review by lossyrob — APPROVED** [review]

**Pull request merged**